File: spring-data-requery-java/src/main/java/com/coupang/springframework/data/requery/configs/AbstractRequeryConfiguration.java
package com.coupang.springframework.data.requery.configs;

import com.coupang.kotlinx.data.requery.listeners.LogbackListener;
import com.coupang.springframework.data.requery.core.RequeryOperations;
import com.coupang.springframework.data.requery.core.RequeryTemplate;
import com.coupang.springframework.data.requery.core.RequeryTransactionManager;
import com.coupang.springframework.data.requery.mapping.RequeryMappingContext;
import io.requery.cache.EmptyEntityCache;
import io.requery.meta.EntityModel;
import io.requery.sql.ConfigurationBuilder;
import io.requery.sql.EntityDataStore;
import io.requery.sql.SchemaModifier;
import io.requery.sql.TableCreationMode;
import lombok.extern.slf4j.Slf4j;
import org.springframework.context.ApplicationContext;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.transaction.PlatformTransactionManager;

import javax.annotation.PostConstruct;
import javax.inject.Inject;
import javax.sql.DataSource;

/**
 * Spring 용 Requery 환경설정 파일입니다.
 *
 * @author debop@coupang.com
 * @since 18. 6. 4
 */
@Slf4j
@Configuration
public abstract class AbstractRequeryConfiguration {

    @Inject
    ApplicationContext applicationContext;

    @Inject
    DataSource dataSource;

    /**
     * Requery용 EntityModel 을 지정해주셔야 합니다. 기본적으로 Models.DEFAULT 를 지정해주시면 됩니다.
     */
    @Bean
    abstract public EntityModel getEntityModel();

    /**
     * Schema Generation 옵션입니다. 기본적으로 {@link TableCreationMode#CREATE_NOT_EXISTS} 을 사용합니다.
     *
     * @return {@link TableCreationMode} 값
     */
    public TableCreationMode getTableCreationMode() {
        return TableCreationMode.CREATE_NOT_EXISTS;
    }

    @Bean
    public io.requery.sql.Configuration requeryConfiguration() {
        return new ConfigurationBuilder(dataSource, getEntityModel())
            // .useDefaultLogging()
            .setEntityCache(new EmptyEntityCache())
            .setStatementCacheSize(1024)
            .setBatchUpdateSize(100)
            .addStatementListener(new LogbackListener())
            .build();
    }

    @Bean(destroyMethod = "close")
    public EntityDataStore<Object> entityDataStore() {
        log.info("Create EntityDataStore instance.");
        return new EntityDataStore<>(requeryConfiguration());
    }

    @Bean
    public RequeryOperations requeryOperations() {
        log.info("Create RequeryTemplate instance.");
        return new RequeryTemplate(entityDataStore(), requeryMappingContext());
    }

    // TODO: 꼭 필요한 Class 인지 다시 판단해보자.
    @Bean
    public RequeryMappingContext requeryMappingContext() {
        RequeryMappingContext context = new RequeryMappingContext();
        context.setApplicationContext(applicationContext);
        return context;
    }

    @Bean
    public PlatformTransactionManager transactionManager() {
        return new RequeryTransactionManager(entityDataStore(), dataSource);
    }

    /**
     * 사용할 Database에 Requery Entity에 해당하는 Schema 를 생성하는 작업을 수행합니다.
     */
    @PostConstruct
    protected void setupSchema() {
        log.info("Setup Requery Database Schema...");

        SchemaModifier schemaModifier = new SchemaModifier(requeryConfiguration());
        try {
            log.debug(schemaModifier.createTablesString(getTableCreationMode()));
            schemaModifier.createTables(getTableCreationMode());
            log.info("Success to setup database schema!!!");
        } catch (Exception e) {
            log.error("Fail to setup database schema!!!", e);
            throw new RuntimeException(e);
        }
    }
}

File: spring-data-requery-kotlin/src/main/kotlin/org/springframework/data/requery/kotlin/configs/AbstractRequeryConfiguration.kt
package org.springframework.data.requery.kotlin.configs

import io.requery.cache.EmptyEntityCache
import io.requery.meta.EntityModel
import io.requery.sql.ConfigurationBuilder
import io.requery.sql.KotlinEntityDataStore
import io.requery.sql.SchemaModifier
import io.requery.sql.TableCreationMode
import mu.KotlinLogging
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.context.ApplicationContext
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.data.requery.kotlin.core.RequeryOperations
import org.springframework.data.requery.kotlin.core.RequeryTemplate
import org.springframework.data.requery.kotlin.core.RequeryTransactionManager
import org.springframework.data.requery.kotlin.listeners.LogbackListener
import org.springframework.data.requery.kotlin.mapping.RequeryMappingContext
import org.springframework.transaction.PlatformTransactionManager
import javax.annotation.PostConstruct
import javax.sql.DataSource

/**
 * AbstractRequeryConfiguration
 *
 * @author debop@coupang.com
 * @since 18. 7. 2
 */
@Configuration
abstract class AbstractRequeryConfiguration {

    companion object {
        private val log = KotlinLogging.logger { }
    }

    @Bean
    abstract fun getEntityModel(): EntityModel

    fun getTableCreationMode(): TableCreationMode {
        return TableCreationMode.CREATE_NOT_EXISTS
    }

    @Bean
    fun requeryConfiguration(dataSource: DataSource, entityModel: EntityModel): io.requery.sql.Configuration {
        return ConfigurationBuilder(dataSource, entityModel)
            .setEntityCache(EmptyEntityCache())
            .setStatementCacheSize(1024)
            .setBatchUpdateSize(100)
            .addStatementListener(LogbackListener<Any>())
            .build()
    }

    @Bean(destroyMethod = "close")
    fun entityDataStore(configuration: io.requery.sql.Configuration): KotlinEntityDataStore<Any> {
        return KotlinEntityDataStore<Any>(configuration).apply {
            log.info { "Create Requery EntityDataStore instance." }
        }
    }

    @Bean
    fun requeryOperations(entityDataStore: KotlinEntityDataStore<Any>, mappingContext: RequeryMappingContext): RequeryOperations {
        log.info { "Create RequeryOperations instance." }
        return RequeryTemplate(entityDataStore, mappingContext)
    }

    @Bean
    fun requeryMappingContext(applicationContext: ApplicationContext): RequeryMappingContext {
        return RequeryMappingContext().apply {
            setApplicationContext(applicationContext)
        }
    }

    @Bean
    fun transactionManager(entityDataStore: KotlinEntityDataStore<Any>, dataSource: DataSource): PlatformTransactionManager {
        return RequeryTransactionManager(entityDataStore, dataSource)
    }

    @Autowired
    lateinit var configuration: io.requery.sql.Configuration

    /**
     * DB Schema 생성이 필요할 경우, Application 시작 시, 존재하지 않는 TABLE을 생성하도록 해줍니다.
     */
    @PostConstruct
    protected fun buildSchema() {
        log.info { "Create Database Schema..." }
        try {
            val schema = SchemaModifier(configuration)
            log.debug { schema.createTablesString(getTableCreationMode()) }
            schema.createTables(getTableCreationMode())
            log.info { "Success to create database schema" }
        } catch(ignored: Exception) {
            log.error(ignored) { "Fail to creation database schema." }
        }
    }
}